encoding/bool_encode.go

```diff
@@ -1,24 +1,42 @@
 package encoding
 
 import (
-	"errors"
 	"io"
 )
 
 // BoolEncoder encodes bool values.
 type BoolEncoder interface {
 	// Encode encodes a collection of bools and writes the encoded bytes to the writer.
-	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardBoolIterator) error
+	Encode(writer io.Writer, valuesIt ForwardBoolIterator) error
 }
 
 // BoolEnc is a bool encoder.
-type BoolEnc struct{}
+type BoolEnc struct {
+	buf []byte
+	t   []byte
+	f   []byte
+}
 
 // NewBoolEncoder creates a new bool encoder.
-func NewBoolEncoder(writer io.Writer) *BoolEnc { return &BoolEnc{} }
+func NewBoolEncoder() *BoolEnc {
+	return &BoolEnc{
+		t: []byte{1},
+		f: []byte{0},
+	}
+}
 
 // Encode encodes a collection of bools and writes the encoded bytes to the writer.
-func (enc *BoolEnc) Encode(writer io.Writer, values ForwardBoolIterator) error {
-	return errors.New("not implemented")
+func (enc *BoolEnc) Encode(writer io.Writer, valuesIt ForwardBoolIterator) error {
+	return runLengthEncodeBool(writer, &enc.buf, enc.writeBool, valuesIt)
+}
+
+func (enc *BoolEnc) writeBool(writer io.Writer, value bool) error {
+	// Write a whole byte since we won't win much by writing a single bit since
+	// varints are byte packed.
+	writeValue := enc.t
```

Comment:
Take it or leave it, but you don't actually need to store the `t` and `f` as fields to avoid a bit of GC / allocation since you can also do sth like
```
var b [1]byte
if value {
  b[0] = trueByte
}
_, err := writer.Write(b[:])
return err
```